feat: combobox

Jinnie opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

* [x] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

* [ ] Bugfix
* [x] Feature
* [ ] Code style update (formatting, local variables)
* [ ] Refactoring (no functional changes, no api changes)
* [ ] Build related changes
* [ ] CI related changes
* [ ] Documentation content changes
* [ ] clarity.design website / infrastructure changes
* [ ] Other... Please describe:

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #248 

## What is the new behavior?

## Does this PR introduce a breaking change?

* [x] Yes
* [ ] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

hippee-lee commented on src/clr-angular/forms/combobox/providers/option-selection.service.ts:
```diff
@@ -1,53 +1,75 @@
+    }
+  }
+
+  get multi(): boolean {
```
nit: is there any way this can be made the same as the attribute its used on? e.g `multiselectable`

Jinnie replied:
`multiselectable` it will be then

hippee-lee commented on src/clr-angular/forms/combobox/utils/combobox-dom-adapter.service.ts:
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
```
nit: Services should go into the providers directory.

Jinnie replied:
Moved them

hippee-lee commented on src/clr-angular/forms/combobox/utils/combobox-noop-dom-adapter.service.ts:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
```
nit: Services should go into the providers directory.

Jinnie replied:
Not sure why they were originally put there - moving them to providers.

hippee-lee commented on src/clr-angular/forms/combobox/_combobox.clarity.scss:
```diff
@@ -1,54 +1,128 @@
+    @include form-fields-appearance($clr-forms-border-color);
+
+    // override the above mixin for combobox-specific styles
+    background: hsl(201, 10%, 96%); // #F4F5F6
```
For theming purposes I think this should be a variable.

Jinnie replied:
Thanks for this, Matt! It brought to my attention the fact that we were far behind with the Dark Theme. Thanks to you and Colin's quickly provided spec, we now have a working one.
Fixed!

hippee-lee commented on src/clr-angular/forms/combobox/_combobox.clarity.scss:
```diff
@@ -1,54 +1,128 @@
+
+      &:focus,
+      &.clr-focus {
+        background: none !important;
```
Why so `important` here and also, why not `inherit`?

Jinnie replied:
I absolutely have no idea why it is there. Not inside it's parent like this:
```
  //Others
    .clr-combobox-input {
      background: none;
      border-bottom: none;

    //   &:focus,
    //   &.clr-focus {
    //     background: none !important;
    //     border-bottom: none !important;
    //   }
    }
```
Removing. It does not seem to do anything meaningful as I test it.

hippee-lee commented on src/clr-angular/forms/combobox/_combobox.clarity.scss:
```diff
@@ -1,54 +1,128 @@
+
+    .label-combobox-pill {
+      margin: $clr_baselineRem_0_125 $clr_baselineRem_0_125 0 0;
+      background-color: hsl(201, 80%, 95%); // #E8F5fC
```
Is this another place for a themeable variable?
same comment for line #66.

Jinnie replied:
Done!

hippee-lee commented on src/clr-angular/forms/combobox/combobox.html:
```diff
@@ -1,32 +1,70 @@
+      <ng-container [ngTemplateOutlet]="textbox"></ng-container>
+    </ng-container>
+
+    <button clrPopoverOpenCloseButton #trigger type="button" class="clr-combobox-trigger" tabindex="-1">
```
Is there a reason users cannot tab to this button?

Jinnie replied:
You can't focus it in the W3C examples too:
https://www.w3.org/TR/wai-aria-practices/examples/combobox/aria1.0pattern/combobox-autocomplete-list.html

hippee-lee commented on src/clr-angular/forms/combobox/combobox.ts:
```diff
@@ -46,69 +72,228 @@ export function comboboxDomAdapterFactory(platformId: Object) {
+  }
+
+  @HostListener('keydown', ['$event'])
+  onKeyDown(event: KeyboardEvent) {
```
Do we need the event argument here? I don't see it used anywhere.

Jinnie replied:
It's not used, but it is still supplied, I think. Unless the compiler complains, there's no overhead to leave it there, for maintenance purposes - better clarity what context is available in this handler. Or am I wrong? Do we gain anything by not specifying it?

Jinnie replied:
Thinking again about it, this is a common keydown handler. It should be a common knowledge how it's distributed. Removing.

hippee-lee commented on src/clr-angular/forms/combobox/combobox.ts:
```diff
@@ -46,69 +72,228 @@ export function comboboxDomAdapterFactory(platformId: Object) {
+    } else {
+      // in theory, setting this again should not cause errors even though we already set it in constructor,
+      // since the initial call to writeValue (caused by [ngModel] input) should happen after this
+      this.optionSelectionService.selectionModel = new SingleSelectComboboxModel<T>();
```
If I read this correct, this line just ovewrites the selection model with a new object that has the same methods. Why not do nothing here?

Jinnie replied:
Well, if I read the code and comments right (this was inherited code to me):
- if we don't specify the `clrMulti` parameter - we set model to single in the constructor
- if there is a specified parameter - we overwrite it with either multi or single
- we overwrite the single above to be able to handle dynamic changes multi/single; which is highly unlikely to happen in practice, but we still do not want to fail if someone does it.

hippee-lee commented on src/clr-angular/forms/combobox/combobox.ts:
```diff
@@ -46,69 +72,228 @@ export function comboboxDomAdapterFactory(platformId: Object) {
+    content: ClrAlignment.START,
+  };
+
+  loadingStateChange(state: ClrLoadingState): void {
```
I'm not seeing this called anywhere. Am I missing something?

Jinnie replied:
Combobox implements LoadingListener. We use this in the async case where we are waiting for options to load after opening the menu.

hippee-lee commented on src/clr-angular/forms/combobox/combobox.ts:
```diff
@@ -46,69 +72,228 @@ export function comboboxDomAdapterFactory(platformId: Object) {
+    this.onTouchedCallback = onTouched;
   }
 
+  setDisabledState(isDisabled: boolean): void {}
```
I'm not sure why this method is here. Its not called and doesn't seem to be part of the WrappedFormCOntrol extensaion.

Jinnie replied:
It's part of the `ControlValueAccessor` interface that we implement, so that we can pretend to be a natural form element.

hippee-lee commented on src/clr-angular/forms/combobox/combobox-container.ts:
```diff
@@ -0,0 +1,93 @@
+    </div>
+  `,
+  // needed for the proper calculation of popover offset that we do to keep label always visible
+  styles: [':host { align-items: stretch; }'],
```
I _think_ this should be in the scss file and targeted to the `clr-combobox-container` element.

Jinnie replied:
Done!

hippee-lee commented on src/clr-angular/forms/combobox/option-items.ts:
```diff
@@ -0,0 +1,98 @@
+      })
+    );
+  }
+
```
I'm concerned that this code will run alot and for large arrays of items could be expensive for apps to run. It looks like this was patterned after the `*clrDgItems` smart iterator in the datagrid and one of the reasons it works is that the only time code is executed is when the differ has changes for the items provided to it. 

Here, we are creating filteredItems everytime change detection is run. Given that this is an input its alot and this is different from the datagrid apporach too, its only an input on the row so change detection isn't nearly as chatty as it is with all the events that trigger ngDoCheck when I run this code now. 

What about this, move lines #68-#82 into a new private function. In my test I called it `updateFilteredItems(): void {...}`. Then I only need to call it in the setter for `@Input('clrOptionItemsOf')` and in the `optionService.inputChanged` subscription. 

I think this limits the work that needs to be done when ngDoCheck is executed.

Jinnie replied:
Tested your suggestion. It really makes difference. Thanks!
I will only leave line 82 (the call to `iterableProxy.ngDoCheck()`) back, so that I'll not be breaking the natural `ngForOf` chain, which I am building upon in this directive.

hippee-lee commented on src/clr-angular/forms/combobox/option.ts:
```diff
@@ -1,35 +1,60 @@
+    this._selected = !!val;
+  }
+
+  @HostBinding('class.active')
```
For consistency, could this be with the other values in the `host:` object?

Jinnie replied:
From what we've previously talked in the team, only static styles go in the `host:` component configuration, while dynamic values we prefer to add as `HostBinding`s.
I don't fully remember the rationale behind it, was it about some productional compilation, AOT, or just preferred style, but I'm sure we didn't want to have dynamic bindings up in the `host:` config.

Jinnie replied:
Living example!
By your request I changed `selectionService.multi` to `selectionService.multiselectable`.
But inside `options.ts` we have this inside the `host:` section
```'[attr.aria-multiselectable]': 'optionSelectionService.multiselectable'```
No refactoring caught it, no IDE caught it, it only appeared during build time.

hippee-lee commented on src/clr-angular/forms/combobox/option.ts:
```diff
@@ -38,36 +63,53 @@ export class ClrOption<T> implements OnDestroy {
-        this.selected = false;
-      }
-    });
+    this.subs.push(
```
nit: I don't think we have a coding guidline for this but generally I put these subscription pushes into the constructor. It looks like it was laready like this. If you agree consider moving the subscription and array push into the constructor and removing this method.

Jinnie replied:
I inherited this code. It's quite common for most of the subscriptions I inherited. We'd better talk if we want to make a convention for this, or leave it to the developer. I guess this was done to keep the code tidy and the constructor small. I personally don't mind either way.

hippee-lee commented on src/clr-angular/forms/combobox/option.ts:
```diff
@@ -38,36 +63,53 @@ export class ClrOption<T> implements OnDestroy {
+    this.updateSelected();
+
     this.toggleService.open = false;
+    // if (!this.optionSelectionService.multi) {
```
Please remove commented code.

Jinnie replied:
Thanks, I forgot removing these. 
Done!

hippee-lee commented on src/clr-angular/forms/combobox/model/multi-select-combobox.model.ts:
```diff
@@ -0,0 +1,89 @@
+
+import { ComboboxModel } from './combobox.model';
+
+export class MultiSelectComboboxModel<T> implements ComboboxModel<T> {
```
I know this has been a huge project but I would like to see spec files for the model functionality like we have for the tree node component models.

Jinnie replied:
It was the first spec I added after this PR was launched :) 
Done.

hippee-lee commented on src/clr-angular/forms/combobox/model/single-select-combobox.model.ts:
```diff
@@ -0,0 +1,37 @@
+
+import { ComboboxModel } from './combobox.model';
+
+export class SingleSelectComboboxModel<T> implements ComboboxModel<T> {
```
I know this has been a huge project but I would like to see spec files for the model functionality like we have for the tree node component models.

Jinnie replied:
It was the first spec I added after this PR was launched :) 
Done.

Jinnie commented on src/clr-angular/forms/combobox/combobox-container.ts:
```diff
@@ -0,0 +1,93 @@
+})
+export class ClrComboboxContainer implements DynamicWrapper, OnDestroy {
+  private subscriptions: Subscription[] = [];
+  @ContentChild(ClrLabel, { static: false })
```
Thanks! Removed.

Jinnie commented on src/clr-angular/forms/combobox/combobox-container.ts:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  addGrid() {
+    if (this.layoutService && !this.layoutService.isVertical()) {
```
Sure. Updated.

Jinnie replied:
Pre-commit hooks failed on this one. Simplified it a bit, but dropped the layoutService?.isV... syntax.

Jinnie commented on src/clr-angular/forms/combobox/combobox-container.ts:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  ngOnDestroy() {
+    if (this.subscriptions) {
```
Absolutely! Probably a residue from a single subscription initial version.

Jinnie commented on src/clr-angular/forms/combobox/combobox.ts:
```diff
@@ -46,69 +72,228 @@ export function comboboxDomAdapterFactory(platformId: Object) {
+  implements ControlValueAccessor, LoadingListener {
+  @ViewChild('comboboxContainer', { static: true })
+  private comboboxContainer: ElementRef;
+  @ViewChild('textboxInput', { static: false })
```
Removed!

Jinnie commented on src/clr-angular/forms/combobox/combobox.ts:
```diff
@@ -46,69 +72,228 @@ export function comboboxDomAdapterFactory(platformId: Object) {
+  }
+
+  private get disabled() {
+    return this.control && this.control.disabled;
```
fixed

Jinnie replied:
Pre-commit hooks failed, dropped this change.

Jinnie commented on src/clr-angular/forms/combobox/combobox.ts:
```diff
@@ -46,69 +72,228 @@ export function comboboxDomAdapterFactory(platformId: Object) {
+    protected renderer: Renderer2,
+    protected el: ElementRef,
+    public optionSelectionService: OptionSelectionService<T>,
+    public cdr: ChangeDetectorRef,
```
No, we're not. Removing!

Jinnie commented on src/clr-angular/forms/combobox/filter-highlight.directive.ts:
```diff
@@ -0,0 +1,39 @@
+
+@Directive({ selector: '[clrFilterHighlight]' })
+export class ClrFilterHighlight<T> implements AfterViewInit, OnDestroy {
+  private subs: Subscription[] = [];
```
We've used this style for subscriptions on several place throughout the code, but I agree, we'd better stick to the convention. Fixed.

Jinnie commented on src/clr-angular/forms/combobox/model/multi-select-combobox.model.ts:
```diff
@@ -0,0 +1,89 @@
+    return item;
+  }
+
+  toString(displayField?: string, index: number = -1): string {
```
I added very detailed unit test to this one. It was not there on your first review, though.

Jinnie commented on src/clr-angular/forms/combobox/model/multi-select-combobox.model.ts:
```diff
@@ -0,0 +1,89 @@
+import { ComboboxModel } from './combobox.model';
+
+export class MultiSelectComboboxModel<T> implements ComboboxModel<T> {
+  model: T[];
```
Sound very good at first sight, but we need at some points to have the model be `null` for validation purposes. So, the checks should stay anyway.
At this point, I don't want to dig into _why_ it needs to be null for validation, but I guess it's a valid requirement, as null or undefined could mean invalid, while [] could be empty but valid.

Jinnie commented on src/clr-angular/forms/combobox/options.ts:
```diff
@@ -1,59 +1,114 @@
+  selector: 'clr-options',
+  template: `
+    <div *ngIf="optionSelectionService.loading" class="clr-options-loading">
+      <span class="spinner spinner-inline">
```
Yes, changed to component. Thanks

Jinnie commented on src/clr-core/common/services/common-strings.default.ts:
```diff
@@ -69,6 +69,10 @@ export const commonStringsDefault: ClrCommonStrings = {
   verticalNavGroupToggle: 'Toggle vertical navigation group',
+  verticalNavToggle: 'Toggle vertical navigation',
+  // combobox
+  delete: 'Delete',
```
Completely agree! Changed to "Delete selected option"

Jinnie commented on src/dev/src/app/combobox/combobox.demo.module.ts:
```diff
@@ -1,22 +1,20 @@
-// Commenting out temporarily. AOT builds do not work when importing outside of scoped library.
-// Can reimport once combobox is added as a public export to @clr/angular
-// import { ClrComboboxModule } from '../../../../clr-angular/forms/combobox/combobox.module';
 
```
Removed this comment. The ClrComboboxModule is in `@clr/angular` now.

Jinnie commented on src/clr-angular/forms/combobox/providers/combobox-focus-handler.service.ts:
```diff
@@ -0,0 +1,233 @@
+@Injectable()
+export class ComboboxFocusHandler implements FocusableItem {
+  constructor(
+    @Inject(UNIQUE_ID) public id: string,
```
Yes, we don't _need_ it to be public. It's a meaningful piece of information on its own and might make sense to have it exposed as a part of the Component's TS API, but as we don't have a specific use case for it, I'd gladly make it private for the initial release.
Done!

Jinnie replied:
Oops! Compilation failed. It comes from the Focusable item implementation. It's public there, so needs to be public here too.

Jinnie commented on src/clr-angular/forms/combobox/providers/combobox-focus-handler.service.ts:
```diff
@@ -0,0 +1,233 @@
+    }
+  }
+
+  // activate() {
```
Done!

Jinnie commented on src/clr-angular/forms/combobox/providers/combobox-focus-handler.service.ts:
```diff
@@ -0,0 +1,233 @@
+        this.focusService.reset(this);
+        // But we only actively focus the textInput when closing, not on initialization.
+        if (this.focusBackOnTextInput) {
+          // if (this.focusBackOnTextInput && document.activeElement.tagName.toLowerCase() === 'body') {
```
Done already :)

Jinnie commented on src/clr-angular/forms/combobox/providers/combobox-focus-handler.service.ts:
```diff
@@ -0,0 +1,233 @@
+  /**
+   * Focus back on the textInput when combobox becomes closed
+   */
+  handleFocusOnClose() {
```
Makes sense.
`handleFocusSubscription()` seems both concise and meaningful.
Made it private and renamed it.

Jinnie commented on src/clr-angular/forms/combobox/providers/combobox-focus-handler.service.ts:
```diff
@@ -0,0 +1,233 @@
+    // todo: remove following todos and comments after verifying it works in various browsers
+    // todo: check why the timeout
+    // looks like it works without it fine
+    // setTimeout(() => {
```
All of them can be removed. Cleaned them up, thanks!

Jinnie commented on src/clr-angular/forms/combobox/providers/combobox-focus-handler.service.ts:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  // this service is only interested in keys that may move the focus
+  private handleTextInput(event) {
```
Definitely! Changed.

Jinnie commented on src/clr-angular/forms/combobox/providers/combobox-focus-handler.service.ts:
```diff
@@ -0,0 +1,233 @@
+          }
+          break;
+        case UP_ARROW:
+          event.preventDefault();
```
Prevent viewport scrolling.
Moved it to a private helper method:
```
  private preventViewportScrolling(event: KeyboardEvent) {...

```

Jinnie commented on src/clr-angular/forms/combobox/providers/combobox-focus-handler.service.ts:
```diff
@@ -0,0 +1,233 @@
+    let preventDefault = false;
+    if (event) {
+      switch (event.keyCode) {
+        case ENTER:
```
I can tell why it happens:
Because as you hit enter and type, no actual selection is happening. Only filtering. You need to go to the list of options (even if only 1 option remains) and make a selection there.
I also don't like that if you type, go to the list and return, your previously typed letters are discarded. It's not unusable, but it's not very usable too. I'm leaving it open for discussion and thinking a bit more on it. The option, as I see them are like this:
- If only one option remains after filtering, allow selecting it with a single Enter-keypress.
- Try to keep the text if you go back form the list to the input.

Jinnie replied:
Looking again at the W3C examples it should close on enter, but not clear the currently typed letters. We can improve on this... I'll handle it after finishing the rest of the review items, as it's severity seems rather small for its scope to me.

Jinnie replied:
Unresolved it, so I can keep track of it. May spawn a low priority issue and will re-resolve it again.

Jinnie replied:
Updated to W3C examples with last commit.

Jinnie commented on src/clr-angular/forms/combobox/providers/combobox-focus-handler.service.ts:
```diff
@@ -0,0 +1,233 @@
+      // we clear out any existing focus on the items
+      this.children.pipe(take(1)).subscribe(items => items.forEach(item => item.blur()));
+
+      // We let the user move focus to where the want, we don't force the focus back on the textInput
```
Fixed, thanks!

Jinnie commented on src/clr-angular/forms/combobox/providers/combobox-offset.service.ts:
```diff
@@ -0,0 +1,23 @@
+  }
+
+  private _labelOffsetChange: BehaviorSubject<number> = new BehaviorSubject(this._labelOffset);
+  public labelOffsetChange: Observable<number> = this._labelOffsetChange.asObservable();
```
You are right, we're not subscribing to it. But I'd like to keep it for the sake of service completeness. It gives us utility, which can be use if we decide to support dynamic page-size changes, what we don't do now.
What are your thoughts about it, and currently unused utility functions in general?

hippee-lee replied:
imo - if its unused it should be removed. My thinking is thus: if there is a use case that needs it, that would be the natural time to add it. I know its only one line, but ... it isn't used and this is an internal service so adding it when its needed is easy. 

However, if you feel strongly about keeping it, I won't block on this 😉

Jinnie replied:
👍 done

Jinnie commented on src/clr-angular/forms/combobox/providers/option-selection.service.ts:
```diff
@@ -1,53 +1,75 @@
-  private currentValue: T;
-  private currentOption: ClrOption<T>;
+  public selectionModel: SelectionModel<T>;
+  public currentOption: ClrOption<T>;
```
I don't understand you very well. Add test? If so, what do you expect it to do? Set and check if it's as expected, so we protect against future changes in implementation?
Please add some more detail about the problem you see here and the suggested changes to handle it.

Jinnie replied:
Done!

Jinnie commented on src/clr-angular/forms/combobox/filter-highlight.directive.ts:
```diff
@@ -0,0 +1,39 @@
+  private filter: string = '';
+  constructor(private element: ElementRef, private optionSelectionService: OptionSelectionService<T>) {}
+
+  ngAfterViewInit() {
```
Definitely yes, thanks!
Added to the code.

Jinnie commented on src/clr-angular/forms/combobox/filter-highlight.directive.ts:
```diff
@@ -0,0 +1,39 @@
+      this.element.nativeElement.innerHTML = this.initialHtml.replace(regex, `<b>$1</b>`);
+    } else {
+      this.element.nativeElement.innerHTML = this.initialHtml;
+    }
```
You are right, it's not needed. I think I needed this some time ago, before resolving some async issue, but now it's obsolete.
Removed it.
The platform check is handled in the previous block, where we don't ever do subscription with this private handler if it's the wrong platform.

bdryanovski commented on src/clr-angular/forms/combobox/model/multi-select-combobox.model.ts:
```diff
@@ -0,0 +1,88 @@
+      this.model.push(item);
+    }
+  }
+
```
```ts
private addItem(item: T) {
  // Check to see that is not already there if not
  if (!this.containsItem(item)) {
    // integrate the first if here, make sure that this.model will be array and add the new item into it.
    this.model = (this.model || []).push(item);
  }
}

Jinnie replied:
I like it. Though you can't combine the assignment with the .push() in one line.
I'll still use it. Calling the this.containsItem() helps it be more concise and clear to understand. Thanks

bdryanovski commented on src/clr-angular/forms/combobox/option-items.ts:
```diff
@@ -0,0 +1,98 @@
+
+  @Input('clrOptionItemsOf')
+  public set rawItems(items: T[]) {
+    this._rawItems = items ? items : [];
```
same as 

```ts
this._rawItems = items || [];
```

Jinnie replied:
I agree, but I find it's a matter of style with no clear advantage of one over the other. What one lacks in bytes, makes up as brevity.

bdryanovski commented on src/clr-angular/forms/combobox/option.ts:
```diff
@@ -38,36 +63,48 @@ export class ClrOption<T> implements OnDestroy {
-      }
-    });
+    this.subscriptions.push(
+      this.optionSelectionService.selectionChanged.subscribe((value: ComboboxModel<T>) => {
```
value is unused

Jinnie replied:
Yes, remained after some refactoring. Thanks, removed it.

bdryanovski commented on src/clr-angular/forms/combobox/option.ts:
```diff
@@ -38,36 +63,48 @@ export class ClrOption<T> implements OnDestroy {
+    const selection = this.optionSelectionService.selectionModel;
+    // Check for null and undefined needed because if the user doesn't assign a value to the option,
+    // all options should not be selected as the value would be null or undefined
+    if (!selection || selection.model === null || selection.model === undefined) {
```
```ts
if (!selection || [null, undefined].includes(selection.model)) { ... }
```

Jinnie replied:
Some inherited code here. It's a matter of style, I think. I still prefer the previous, as it is easier to read. Especially about `null` and `undefined`.
I'd personally do it like `!selection || selection.model`, but the more explicit check does not bother me too, and I'd prefer not to refactor it unless I see a clear advantage.

bdryanovski commented on src/clr-angular/main.scss:
```diff
@@ -9,5 +9,6 @@
 
 // OG Theme: This is for generating the OG Clarity Theme
 
+//@import './utils/theme.dark.clarity';
```
... do we need this ?

Jinnie replied:
oops! thanks, it's forgotten. This was for hard-wired testing of the component with the dark theme. Removing.

Jinnie commented on src/clr-angular/forms/combobox/model/multi-select-combobox.model.ts:
```diff
@@ -0,0 +1,88 @@
+  model: T[];
+
+  containsItem(item: T): boolean {
+    return this.model ? this.model.indexOf(item) > -1 : false;
```
Your suggestion is nicely structured, but harder to convey the intention. I'll replace the indexOf with includes, as we agreed it gets transpiled well for old browsers.
Thanks.

Jinnie commented on src/clr-angular/forms/combobox/model/single-select-combobox.model.ts:
```diff
@@ -0,0 +1,40 @@
+  setSelected(item: T, selected: boolean): void {
+    if (selected) {
+      this.model = item;
+    } else if (item === this.model) {
```
Yes, thanks!
Done

Jinnie commented on src/clr-angular/forms/combobox/model/multi-select-combobox.model.ts:
```diff
@@ -0,0 +1,88 @@
+    return this.model ? this.model.indexOf(item) > -1 : false;
+  }
+
+  setSelected(item: T, selected: boolean): void {
```
You are absolutely right. This setSelected with boolean option was bugging me a lot. 
Toggle is also not the right term. So I changed it to a pair of "select" and "unselect" methods. This split all the old "setSelected" implementations (which you commented too below) into more manageable and easy to understand smaller methods.
Thanks!

Jinnie commented on src/clr-angular/forms/combobox/_combobox.clarity.scss:
```diff
@@ -6,49 +6,121 @@
 @include exports('combobox.clarity') {
-  .clr-combobox {
-    display: inline-block;
+  .clr-aria-only {
```
Thanks! `.clr-sr-only` is the correct one.
Changed and removed the `.clr-aria-only` from the CSS file.

hippee-lee commented:
> I don't understand you very well. Add test? If so, what do you expect it to do? Set and check if it's as expected, so we protect against future changes in implementation?
Please add some more detail about the problem you see here and the suggested changes to handle it. 

Yes, I was thinking of a simple coverage test. Since it is being set/get externally it would be nice to know if the implementation changes if it were to break that contract.

rbirkgit commented:
Just tested a little but found some issues. This is on the "My State" sample:

* Typing "flo" I see options. I can cursor down to option, but no way to go back to text input
* After selecting an option, I can now only type one letter. 
* Pressing escape does not bring you back to input either. Have to press tab and shift-tab to go back. May be same bug as first above.

Jinnie commented:
> Just tested a little but found some issues. This is on the "My State" sample:
> 
> * Typing "flo" I see options. I can cursor down to option, but no way to go back to text input
> * After selecting an option, I can now only type one letter.
> * Pressing escape does not bring you back to input either. Have to press tab and shift-tab to go back. May be same bug as first above.
> 
> With the tags/labels. The text is not centered vertically (may just be the sample and not the control). It would be nice though to be able to press left cursor / backspace key to select / remove a tag. I find it extra work to use the tab key and space. This is how I'm used to the tags to work.

Thanks for the check and report, @rbirkgit ! We indeed need to polish the input/menu/input navigation behavior a bit more. Btw, backspace deletes last tags too.
I intend to have one more iteration over the keyboard and focus navigation, in regard to the W3C ARIA specs and examples. Your notes will help!

Jinnie commented on src/clr-angular/forms/combobox/providers/option-selection.service.spec.ts:
```diff
@@ -1,104 +1,79 @@
-      fakeOption1 = new ClrOption(new ClrPopoverToggleService(), new ElementRef(null), null, optionSelectionService);
-      fakeOption2 = new ClrOption(new ClrPopoverToggleService(), new ElementRef(null), null, optionSelectionService);
+      optionSelectionService.selectionModel = <ComboboxModel<string>>new SingleSelectComboboxModel<string>();
     });
```
I don't think we have a convention about this. I've seen big numbers of both throughout the code. Or I even think I've seen using `as` quite less often. And in this specific case I even prefer the assignment to explicitly fail if something wrong is being cast, while `as` will silently swallow the exception.

Jinnie commented on src/clr-angular/forms/combobox/providers/option-selection.service.ts:
```diff
@@ -1,53 +1,73 @@
 
 @Injectable()
 export class OptionSelectionService<T> {
-  private currentValue: T;
```
The models themselves are tested in their respective spec. Here we just need an instance of ComboboxModel, no matter which one. They both implement their parent API, and from the OptionSelectionService point of view, they are the same - just a ComboboxModel.
We don't need to test the models twice.

Jinnie commented on src/clr-angular/forms/combobox/providers/option-selection.service.ts:
```diff
@@ -1,53 +1,73 @@
-  private currentOption: ClrOption<T>;
+  public selectionModel: ComboboxModel<T>;
+  public currentOption: ClrOption<T>;
+  public loading: boolean = false;
```
As we add code that can be used in future when it is required to be used (ref. the observable in the label offset service), won't it similarly make sense to add this test when the implementation is actually changed?
Test added. 😉

Jinnie commented on src/clr-angular/forms/combobox/providers/option-selection.service.ts:
```diff
@@ -1,53 +1,73 @@
 @Injectable()
 export class OptionSelectionService<T> {
-  private currentValue: T;
-  private currentOption: ClrOption<T>;
```
Yes! Thanks. One test down, too.
Removed

Jinnie commented on src/clr-angular/forms/combobox/providers/option-selection.service.spec.ts:
```diff
@@ -1,104 +1,79 @@
 
     it('provides an observable to subscribe to change in value', () => {
-      expect(optionSelectionService.valueChanged).toBeDefined();
+      expect(optionSelectionService.inputChanged).toBeDefined();
```
Won't ignore, thanks!
Changed.

Jinnie commented:
Last commit finalizes ARIA effort to use aria-activedescendant on the popover and roving tabindex for the options. The pattern is in sync with W3C examples for both combobox and grid-layout pills component. Text-input is now always focused and editable.
Before we remove the DO NOT MERGE tag the following needs to be completed:
- Add tests for the new (post-aria-activedescendant) classes and scenarios. Old tests are all fixed
- Add documentation

Open again for reviews, keeping in mind that several tests are still missing.

Jinnie commented:
**ARIA examples** used as basis:
- https://www.w3.org/TR/wai-aria-practices/examples/combobox/aria1.1pattern/listbox-combo.html (Example 1: List Autocomplete with Manual Selection)
- https://www.w3.org/TR/wai-aria-practices/examples/grid/LayoutGrids.html (Example 2: Pill List...)

**Known issues**:
  - Does not close popover on trigger click
    - Not optimal, but there are W3C examples with this behavior (see "Example 3: List with Inline Autocomplete" in the list-combobox example above).
  - May be good to keep the popover open on Enter press with multiselect!?
- Popover issues in IE <- under investigation.

gnomeontherun commented on src/clr-angular/forms/combobox/options.ts:
```diff
@@ -1,59 +1,140 @@
+  }
+
+  ngAfterViewInit() {
+    this.focusHandler.listbox = this.el.nativeElement;
```
Need a platform check

Jinnie replied:
Not yet. It's just assignment of a nativeElement. We only apply any function on it at `ComboboxFocusHandler.addFocusOnBlurListener()` and there we have the platform check.
We are transferring nativeElements without platform check at many places in the code.

gnomeontherun replied:
I'm more worried about the subscription below, though it should never fire a scroll event in SSR.

gnomeontherun commented on src/clr-angular/forms/combobox/_combobox.clarity.scss:
```diff
@@ -1,69 +1,188 @@
-    transform: translateX(-1 * $clr-ng-select-trigger-dimensions);
   }
 
   .clr-options {
```
Would like to make sure we prefix classes fully, so `.clr-combobox-options` and update this throughout the CSS.

gnomeontherun commented on src/clr-angular/forms/combobox/_combobox.clarity.scss:
```diff
@@ -1,69 +1,188 @@
+    }
+  }
+
+  [clrFilterHighlight] b {
```
We should have the filter highlight render a class instead so we can match on it instead of matching on the attribute.

Jinnie replied:
Done.

gnomeontherun commented on src/clr-angular/forms/combobox/_combobox.clarity.scss:
```diff
@@ -1,69 +1,188 @@
+      color: $clr-ng-select-pill-font-color;
+    }
+
+    .label-combobox-pill {
```
I wonder if we can make this generic and allow labels to have this outside of the combobox as well?

gnomeontherun commented on src/clr-angular/forms/combobox/_properties.combobox.scss:
```diff
@@ -7,6 +7,11 @@
     :root {
       --clr-ng-select-trigger-color: var(--clr-color-neutral-500);
       --clr-ng-select-font-size: #{$clr-p2-font-size};
+      --clr-ng-select-input-background: #{$clr-ng-select-input-background-color};
```
Update prefixes throughout to be `--clr-combobox` and not `--clr-ng-select`

Jinnie replied:
Done

gnomeontherun commented on src/clr-angular/forms/combobox/combobox.ts:
```diff
@@ -45,70 +62,288 @@ export function comboboxDomAdapterFactory(platformId: Record<string, any>) {
+  @ViewChild('trigger') trigger: ElementRef;
+  @ContentChild(ClrOptionSelected) optionSelected: ClrOptionSelected<T>;
+
+  private onChangeCallback: any;
```
Can probably type with `() => any`?

Jinnie replied:
Done. Also for onTouchedCallback.

gnomeontherun commented on src/clr-angular/forms/combobox/filter-highlight.directive.ts:
```diff
@@ -0,0 +1,57 @@
+    }
+    const regex = new RegExp(`(${this.sanitizeForRegexp(this.filter)})`, 'gi');
+    if (this.filter && this.initialHtml.match(regex)) {
+      this.element.nativeElement.innerHTML = this.initialHtml.replace(regex, `<b>$1</b>`);
```
We should not be doing this, as it has a potential to be a vector for hijacking a script if we are doing something wrong with our sanitation and DOM manipulation.

If we just want to bold text, we should just add a CSS class that sets the font-weight.

Jinnie replied:
We want to be highlighting a substring/s of the whole string. That's the problem. 
Like "**Am**erican S**am**oa" (for search string "am")
We have to drop the whole highlighting scenario if we think it's too risky. I don't mind, it will make the API simpler too. But we have it in the current design specs.

gnomeontherun replied:
Let's review this together sometime this week and make sure we can do this correctly.

gnomeontherun commented on src/clr-angular/forms/combobox/options.ts:
```diff
@@ -1,59 +1,140 @@
+    <!-- Rendered if data set is empty -->
+    <div *ngIf="emptyOptions">
+      <span class="clr-options-empty-text">
+        No results
```
This is not localized, and can user pass this into the component somehow?

Jinnie replied:
Of course it should be. Thanks.
Done.

Jinnie commented on tests/tslint/dist/noBarrelImportsRule.js:
```diff
@@ -0,0 +1,85 @@
+"use strict";
```
Oops, got in during the merge. I remember seeing it. Probably my mistake adding it back.
Removing.

Jinnie commented on src/website/src/settings/componentlist.json:
```diff
@@ -101,6 +101,16 @@
+      "ui": 20,
+      "core": -1,
+      "ng": 20,
+      "dox": 20
```
Now that's probably after an unsuccessful git reset, when I had to manually restore parts of lost code.
Thanks for catching it!
Fixed

Jinnie commented on src/clr-angular/forms/combobox/_properties.combobox.scss:
```diff
@@ -7,6 +7,11 @@
+      --clr-ng-select-input-background: #{$clr-ng-select-input-background-color};
+      --clr-ng-select-pill-background-color: #{$clr-ng-select-pill-background-color};
+      --clr-ng-select-pill-border-color: #{$clr-ng-select-pill-border-color};
+      --clr-ng-select-pill-font-color: #{$clr-ng-select-pill-font-color};
```
Done

Jinnie commented on src/clr-angular/forms/combobox/_variables.combobox.scss:
```diff
@@ -1,13 +1,20 @@
 // This software is released under MIT license.
 // The full license information can be found in LICENSE in the root directory of this project.
 
 $clr-ng-select-min-height: $clr_baselineRem_1 !default;
```
Done

Jinnie commented on src/clr-angular/forms/combobox/combobox-container.ts:
```diff
@@ -0,0 +1,87 @@
+  },
+  providers: [IfErrorService, NgControlService, ControlIdService, ControlClassService, ComboboxOffsetService],
+})
+export class ClrComboboxContainer implements DynamicWrapper, OnDestroy {
```
Good direction. Thanks, done.

Jinnie commented on src/clr-angular/forms/combobox/combobox-container.ts:
```diff
@@ -0,0 +1,87 @@
+
+  ngAfterViewInit() {
+    this.offsetService.labelOffset =
+      this.controlContainer.nativeElement.offsetHeight - this.el.nativeElement.offsetHeight;
```
This and the controlContainer ViewChild. This comment helped.

Jinnie commented on src/clr-angular/forms/combobox/combobox.html:
```diff
@@ -1,32 +1,100 @@
-    </button>
+<!-- The (click) handler is needed to auto-focus on input field which can not currently occupy the whole
+     width of the component, after being wrapped to a new line -->
+<div class="clr-combobox-container" clrPopoverAnchor (click)="focusInput()"
```
True. `clr-combobox-wrapper` is better.
Done.

Jinnie commented on src/dev/src/app/combobox/combobox.demo.ts:
```diff
@@ -1,9 +1,329 @@
  * The full license information can be found in LICENSE in the root directory of this project.
  */
 import { Component } from '@angular/core';
+import { Observable } from 'rxjs/index';
```
Oops! Fixed.

Jinnie commented on src/clr-angular/forms/combobox/combobox.module.ts:
```diff
@@ -1,22 +1,59 @@
+    ClrKeyFocusModule,
+    ClrCommonFormsModule,
+    ClrPopoverModuleNext,
+    ClrSpinnerModule,
```
Yes, it's missing. But why would we need it? I don't remember using any of the directives there.

gnomeontherun replied:
You use `IF_ACTIVE_ID_PROVIDER` which means if someone were to import just this module for testing they'd possibly get a missing provider.

Jinnie replied:
Added to the imports.

Jinnie commented on src/clr-angular/forms/combobox/combobox.module.ts:
```diff
@@ -1,22 +1,59 @@
+    ClrOptionItems,
+  ],
+  entryComponents: [ClrComboboxContainer],
+  schemas: [NO_ERRORS_SCHEMA],
```
Don't remember why/when I needed the last two, but now all seems to pass without them.
Removed both last two lines.

Jinnie commented on src/clr-angular/forms/combobox/combobox.module.ts:
```diff
@@ -1,22 +1,59 @@
+    ClrSpinnerModule,
+  ],
+  declarations: [
+    ClrCombobox,
```
It's not done in any of the other form components. I preferred to stay compliant with the local style.

Jinnie commented on src/clr-angular/forms/combobox/combobox.module.ts:
```diff
@@ -1,22 +1,59 @@
+    ClrFilterHighlight,
+    ClrOptionItems,
+  ],
+  entryComponents: [ClrComboboxContainer],
```
Removed. Should I remove it from all other form components, too?

Jinnie replied:
Leaving the local change only. Cleaning up from other components may be a separate chore task.

Jinnie commented on src/clr-angular/forms/combobox/model/multi-select-combobox.model.ts:
```diff
@@ -0,0 +1,87 @@
+import { ComboboxModel } from './combobox.model';
+
+export class MultiSelectComboboxModel<T> implements ComboboxModel<T> {
+  model: T[];
```
There are some direct assignments in test, which can be refactored, but the currently _legal_  way to assign it is through the `OptionSelectionService`. Not sure how we can implement it without exposing a setter too.

gnomeontherun replied:
You can have a setter without a getter. I would prefer that the tests use the _legal_ way :).

Jinnie replied:
I'm sorry, but it seems that we have much more cases that need access to the model and I don't see how we can avoid this:
- OptionSelectionService.setSelectionValue() has checks if we are not setting the same model.
- Combobox.ts
   - Multiple model accesses, tied to the ControlValueAccessor implementation
   - User in logic determining display name
   - Used in prepopulation of the search text with the 1st selected item
   - Used in retrieving the id of the active-descendant for the keyboard navigation
- FocusHandler
   - Retrieving the index of the currently focused item
   - ScrollIntoView functionality for currently focused item
   - Retrieving and focusing first selected item in a pre-selected combobox 
If we add both getter and setter - then it's no different from just a plain public property.

gnomeontherun replied:
My main point is that there isn't much value in the rest of this class if you can directly modify the model, as its really just providing helper methods instead of managing the model. If someone sets the model to an empty array instead of a null, then we get into a strange state for validation, as one example. The reason I'm focusing on this is that the model is the most likely place we'll get something out of sync between user interactions.

Jinnie replied:
I understand your point, and yes - currently it is just a model holder with utility methods.
And my point is that, as long as we implement ControlValueAccessor, we're required to allow direct access. The only meaningful action item I can think of, is to introduce setter and getter and force null in the case of empty array in the setter.
If you have any other actionable ideas, please advise!
I also can't really advocate why it is designed as it is, I don't even know who designed it initially. But it's buried deep inside the combobox architecture and if we decide that it's important to do it differently, we should be aware that it may take a bit more time, as it will affect many aspects of the Combobox implementation that we currently have. If we want another architecture, we should talk about it and define its strict requirements.

gnomeontherun replied:
Alright, let's then add a comment about this as a TODO for this.

Jinnie commented on src/clr-angular/forms/combobox/model/multi-select-combobox.model.ts:
```diff
@@ -0,0 +1,87 @@
+  toString(displayField?: string, index = -1): string {
+    let displayString = '';
+
+    if (this.model) {
```
Documented it. I'll have to check if we're at all using it, as there were changes in the A11Y approach used. Leaving open until this check is done.

Jinnie commented on src/clr-angular/utils/key-codes/key-codes.ts:
```diff
@@ -4,11 +4,14 @@
  * The full license information can be found in LICENSE in the root directory of this project.
  */
 
+// TODO Consider deprecation in favor of Core KeyCodes
```
Yes, after the merge I realized this. 
Removed.

Jinnie commented on src/clr-angular/utils/i18n/common-strings.default.ts:
```diff
@@ -69,12 +69,16 @@ export const commonStringsDefault: ClrCommonStrings = {
   timelineStepError: 'Error',
   timelineStepProcessing: 'In progress',
+  // Combobox
+  delete: 'Delete selected option',
```
Done.

Jinnie commented on src/clr-angular/utils/focus/key-focus/roving-tabindex.ts:
```diff
@@ -0,0 +1,89 @@
+import { FocusableItem } from './interfaces';
+
+@Component({
+  selector: '[clrRovingTabindex]',
```
`Roving tabindex` is a named algorithm:
https://www.w3.org/TR/wai-aria-practices/#kbd_roving_tabindex

Jinnie commented on src/clr-angular/utils/focus/key-focus/roving-tabindex.ts:
```diff
@@ -0,0 +1,89 @@
+
+  private setTabindex(item: FocusableItem, value: number) {
+    if (item instanceof HTMLElement) {
+      this.renderer.setAttribute(item, 'tabindex', String(value));
```
Didn't know that typescript adds `toString()` to the simple type `number`.
Okay, changed.

Jinnie commented on src/clr-angular/utils/focus/key-focus/roving-tabindex.ts:
```diff
@@ -0,0 +1,89 @@
+  selector: '[clrRovingTabindex]',
+  template: '<ng-content></ng-content>',
+})
+export class ClrRovingTabindex extends ClrKeyFocus {
```
We removed all this code from the ClrKeyFocus a week before I needed it here. Without it, I'd have to spread tabindex calculation code all throughout the hierarchy of combobox inner components. Now it takes 5 lines of code overall.
Shijir needed KeyFocus without the tabindex. So, we agreed his case is valid and the code was removed. But the Roving Tabindex is a recognized ARIA pattern that we may need on many places and it steps upon the ClrKeyFocus. 
Alternatively, we can just duplicate the code from ClrKeyFocus here. I don't see advantages to it though.

gnomeontherun replied:
As we discussed, there are other approaches but at this point I don't know that we want to make any changes. We just want to have this in mind and long term this might get sorted out.

Jinnie commented on src/clr-angular/utils/focus/focusable-item/basic-focusable-item.service.ts:
```diff
@@ -27,8 +27,15 @@ export class BasicFocusableItem implements FocusableItem {
+
+      // TODO: figure out if i can move this somewhere so it doesn't have to use setTimeout
+      // TODO: optimize to scrollIntoView only if hidden
+      setTimeout(() =>
```
No idea. I inherited this code. My only contribution to it is using "nearest" alignment.
Will try to research for improvement on the next iteration.

gnomeontherun replied:
I'm not sure this is needed, but let me know.

Jinnie replied:
It works perfectly well without this timeout on all supported browsers. 
Though researching this helped me detect and fix a regression with the keyboard navigation in IE11.
Removed.

Jinnie commented on src/dev/src/app/combobox/combobox.demo.html:
```diff
@@ -1,12 +1,248 @@
+
+    <clr-combobox-container>
+      <label>Two</label>
+      <clr-combobox [(ngModel)]="vertical.two" name="two" required [disabled]="disabled" class="dynamic">
```
Legacy. Inherited. I noticed this, but forgot to remove it. Thanks!
Deleted.

Jinnie commented on src/clr-angular/forms/combobox/option-items.directive.ts:
```diff
@@ -0,0 +1,120 @@
+      this.filteredItems = this._rawItems.filter(item => {
+        if (typeof item !== 'object') {
+          return (
+            String(item)
```
Done

hippee-lee commented on src/clr-angular/forms/combobox/option-items.directive.ts:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  ngDoCheck() {
+    this.iterableProxy.ngDoCheck();
```
Is there a reason for calling the iterableProxy.ngDoCheck here instead of on line #101?IIR - the iterableProxy needs the component to manually call ngDoCheck whenever items are changed and that propagates and allows the app to detecct changes. 

I believe when it is called in `updateItems` the ngDoChecks will be picked up here as well and run. But perhaps there is a perfermance issue I'm not seeing.

Jinnie replied:
If I move it to 101 it seems to work, but it does start to fail test cases in the 'correctly mutates and resets an array with trackBy' test.
Also, it seems normal that when you wrap some component, and you redefine some of their lifecycle methods, you should call the basic component's methods from your redefined methods. Otherwise you're messing with the natural lifecycle of the original component.
If you have a feeling that the other approach is better, please try it, check it against the tests and try to prove that it really is better.
I am open for f2f discussion, won't be closing this conversation.

Jinnie replied:
We looked at the code together and found an even better solution - keep the call still in the local ngDoCheck, but inside the block that tests for differences. 
It works with the same efficiency as the suggested by Matt call-from-updateItems scenario, but does not break the trackBy tests.

hippee-lee commented on src/clr-angular/forms/combobox/option-items.directive.spec.ts:
```diff
@@ -0,0 +1,206 @@
+        const updatedContent = this.fixture.elementRef.nativeElement.textContent;
+
+        /* This took me quite some time to research, so it needs a detailed explanation.
+           The data not updating immediately does not mean that the new value will not be added to the combobox.
```
I suspect that this happens because the updateItems method isn't telling the iterableProxt to run ngDoCheck.

Jinnie replied:
Moving the call of ngDoCheck to the updateItems method had no effect on this test.

Jinnie replied:
We talked and agreed that this issue is not related to the way that the directive works but that we're not monitoring for internal changes in the data array. This, so far, is acceptable, as it gets updated on popover reopening and popovers staying consistent while open is actually desired, even if there is a chance to get this data updated.

hippee-lee commented on src/clr-angular/forms/combobox/option-items.directive.spec.ts:
```diff
@@ -0,0 +1,206 @@
+        this.testComponent.numbers.unshift(42);
+        this.fixture.detectChanges();
+        const unshiftedItem = this.fixture.nativeElement.querySelector('li');
+        // expect(this.itemsProvider.displayed).toEqual([42, 1, 2, 3, 4, 5]);
```
Shouldn't it test that the displayed items reflect the new value unshifted into the array, in the correct position? I think thats why I wrote this in this way for the datagrid trackBy tests.

Jinnie replied:
Oops! Didn't see I forgot that commented, as it had to have the itemsProvider service replaced with a new source of truth.
Fixed!

hippee-lee commented on src/clr-angular/forms/combobox/option-items.directive.ts:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  @Input('clrOptionItemsField')
+  public set field(field: string) {
```
I'm not seeing a test for this setter that makes sure the correct value is set in this class and in the optionService.

Jinnie replied:
The service gets checked in the 'sets display field' test.
I'll add the local field check to this test too.
Fixed.

hippee-lee commented on src/clr-angular/forms/combobox/option-items.directive.ts:
```diff
@@ -0,0 +1,120 @@
+  ) {
+    this.iterableProxy = new NgForOf<T>(this.vcr, this.template, this.differs);
+    this.subscriptions.push(
+      optionService.inputChanged.subscribe(filter => {
```
It needs to test that htis subscription is properly fired and the side effects are correct. E/g this.filter is properly updated when the subscription streams a new event and that the updateItems method is fired.

Jinnie replied:
The updateItems method and this.filter property are private. I disagree that we should be testing private code that is implementation specific, as this most often results in inflexible, non-portable tests that break once we decide to change/improve the implementation. 
Let's talk about this.

Jinnie replied:
We agreed on a call that this specific case is covered by functional scenarios and the calls inside the observable handler are private and implementation specific.

hippee-lee commented on src/clr-angular/forms/combobox/options.ts:
```diff
@@ -1,59 +1,140 @@
-export class ClrOptions extends AbstractPopover implements OnDestroy {
-  private sub: Subscription;
+    <!-- Rendered if data set is empty -->
+    <div *ngIf="emptyOptions">
```
I hadn't ever thought of using a QueryList like this, neat!

I think it just needs a test that the correct things get set when a a new ClrOptions component is created (and the gettor is returning the correct thing). 

I would like to see that if there is an *ngIf on one of the options the QueryList rflects the correct ClrOptions. I'm not 100% sure about this as it might just be retesting the QueryList behavior so you your discresion here 

hippee-lee commented on src/clr-angular/forms/combobox/options.ts:
```diff
@@ -1,59 +1,140 @@
-import { Point } from '../../popover/common/popover';
+import {
+  Component,
+  ContentChildren,
```
This needs a `View Basics` test for both the loading state and the loaded state.

Jinnie replied:
Added. Also a check for "No results".

hippee-lee commented on src/clr-angular/forms/combobox/options.ts:
```diff
@@ -1,59 +1,140 @@
+} from '@angular/core';
+import { DOCUMENT } from '@angular/common';
 import { POPOVER_HOST_ANCHOR } from '../../popover/common/popover-host-anchor.token';
+import { OptionSelectionService } from './providers/option-selection.service';
```
I think this was mentioned but just in case this should be a common string so consumers can override it.

Jinnie replied:
This comment is misplaced, but I assume you talk about the "Loading" text.
Localized it. Done.

hippee-lee commented on src/clr-angular/forms/combobox/options.ts:
```diff
@@ -1,59 +1,140 @@
+  Input,
+  Optional,
+  QueryList,
+  AfterViewInit,
```
This needs a View Basics test that this class gets added / removed properly.

hippee-lee commented on src/clr-angular/forms/combobox/options.ts:
```diff
@@ -1,59 +1,140 @@
+  Inject,
+  Input,
+  Optional,
+  QueryList,
```
Needs a View Basics test that ensures this class is always added.

Jinnie replied:
Comment misplaced. If you are talking about `.clr-combobox-options`, we have such test.
If not, please specify which class do you mean.

Jinnie replied:
Confirmed. Fixed.

hippee-lee commented on src/clr-angular/forms/combobox/options.ts:
```diff
@@ -1,59 +1,140 @@
+  AfterViewInit,
+  OnDestroy,
+} from '@angular/core';
+import { DOCUMENT } from '@angular/common';
```
Needs a test that when emptyOptions returns true this is rendered.

Jinnie replied:
Done as part of the loading/searching test above.

hippee-lee commented on src/clr-angular/forms/combobox/options.ts:
```diff
@@ -1,59 +1,140 @@
+  Inject,
+  Input,
+  Optional,
+  QueryList,
```
This needs a `Template API` test (imo - searchText is a function that is called by the template, so its not a View Basics test) that ensures the DOM accurately interpolates the searchText as it gets updated.

Jinnie replied:
Done

hippee-lee commented on src/clr-angular/utils/popover/providers/popover-toggle.service.ts:
```diff
@@ -59,4 +60,14 @@ export class ClrPopoverToggleService {
+
+  private _popoverAligned: Subject<HTMLElement> = new Subject();
+
+  public get popoverAligned(): Observable<HTMLElement> {
```
This needs a test that this public property return an observable if that type.

Jinnie replied:
Done.

hippee-lee commented on src/clr-angular/utils/popover/providers/popover-toggle.service.ts:
```diff
@@ -59,4 +60,14 @@ export class ClrPopoverToggleService {
+    return this._popoverAligned.asObservable();
+  }
+
+  public popoverAlignedEmit(popoverNode: HTMLElement) {
```
Need a test for the behavior when a consumer is subscribed.

Jinnie replied:
Done.

hippee-lee commented on src/clr-angular/forms/combobox/option.ts:
```diff
@@ -3,71 +3,94 @@
-  @HostBinding('class.active') selected = false;
+  private _value: T;
+  @Input('clrValue')
+  set value(value: T) {
```
1. Are we concerned if the consumer does not pass an value?
2. See note about tests on optionProxy property

Jinnie replied:
1. I don't think the consumer should have high expectations if he skips the value. It's part of the API. If he does - he'll get "undefined" value, which is exactly what he requested. Not sure what more we can do about it.
2. Done, as part of the optionProxy tests scenario. Thanks!

Jinnie replied:
We agreed on a call that for now this is enough and there may be risks to break a scenario where the user may really want to use the "undefined" value for one of his items.

hippee-lee commented on src/clr-angular/forms/combobox/option.ts:
```diff
@@ -3,71 +3,94 @@
-  private subscription: Subscription;
+export class ClrOption<T> implements OnInit {
+  // A proxy with only the necessary data to be used for a11y and the focus handler service.
+  public optionProxy: OptionProxy<T> = new OptionProxy(null, null);
```
I think optionProxy needs a couple of tests. I think these shoul dbe in a describe for the `Template API` and the use cases I see for this property are:
1. It instantiates a new OptionProxy when created
2. It sets the correct id on the OptionProxy
3. It sets the correct value on the Option OptionProxy and for optionId/value props
4. It gets the correct value(s) for [id]/[clrValue]
5. It updates the [id]/[clrValue] when there is a binding

Jinnie replied:
Done! Thanks

hippee-lee commented on src/clr-angular/forms/combobox/option.ts:
```diff
@@ -3,71 +3,94 @@
+
+  private _id: string;
+  @Input('id')
+  set optionId(id: string) {
```
1. ~Are we concerned if the consumer does not pass an id?~
2. See note about tests on optionProxy property

Jinnie replied:
1. Without specified id, the user will get one generated from the container.
2. Done, thanks!

hippee-lee commented on src/clr-angular/forms/combobox/option.ts:
```diff
@@ -3,71 +3,94 @@
 
   constructor(
-    private toggleService: ClrPopoverToggleService,
     @Optional()
```
Should this really be optional? We throw an error below if its not available.

Jinnie replied:
Legacy code, removed! Thanks

hippee-lee commented on src/clr-angular/forms/combobox/option.ts:
```diff
@@ -3,71 +3,94 @@
-  // Lifecycle Methods
-  ngOnDestroy(): void {
-    this.subscription.unsubscribe();
+  @HostBinding('class.clr-focus')
```
I don't see a test for the true/false states of this hostbinding. Imo it should be part of the `View Basics` suite

Jinnie replied:
Thanks! Added these tests.

hippee-lee commented on src/clr-angular/forms/combobox/options.ts:
```diff
@@ -1,59 +1,140 @@
+@Component({
+  selector: 'clr-options',
+  template: `
+    <div *ngIf="optionSelectionService.loading" class="clr-options-loading">
```
This needs a View Basics test for both the loading state and the loaded state.

Jinnie replied:
Your IDE probably double-posted this. 
Test added, as commented previously.

hippee-lee commented on src/clr-angular/forms/combobox/options.ts:
```diff
@@ -1,59 +1,140 @@
-export class ClrOptions extends AbstractPopover implements OnDestroy {
-  private sub: Subscription;
+    <!-- Rendered if data set is empty -->
+    <div *ngIf="emptyOptions">
```
Needs a test that when emptyOptions returns true this is rendered.

Jinnie replied:
Actually, we're testing this in isolation (no parent, no clr-option children). So, it is empty. I added a test for this as part of one of your other requests. 
So, we could say - done.

Jinnie replied:
Added some tests with `clr-option` child and tested `emptyOptions` in dependence with real children.

hippee-lee commented on src/clr-angular/forms/combobox/options.ts:
```diff
@@ -1,59 +1,140 @@
+  }
+
+  _items: QueryList<ClrOption<T>>;
+  @ContentChildren(ClrOption)
```
I hadn't ever thought of using a QueryList like this, neat!

I think it just needs a test that the correct things get set when a a new ClrOptions component is created (and the gettor is returning the correct thing).

I would like to see that if there is an *ngIf on one of the options the QueryList rflects the correct ClrOptions. I'm not 100% sure about this as it might just be retesting the QueryList behavior so you your discresion here

Jinnie replied:
I see. Yes, we could test this. But this will break the isolation of the tests we have now and I'll need to import and add clr-option elements to the testing template.
This will complicate the testing scenario, but if you think it will bring value, we could add it?

Jinnie replied:
Added some tests with `clr-option` child.

hippee-lee commented:
Sorry for all of the extra comments. I used VSCode with the Github Pull Requests plugin and some of my comments ended up in the wrong place. I think I found them all but If there is anything that seems out of place  lmk and I'll fix it.

Jinnie commented on src/clr-angular/forms/combobox/option.ts:
```diff
@@ -3,71 +3,94 @@
+  get selected() {
+    return (
+      this.optionSelectionService.selectionModel && this.optionSelectionService.selectionModel.containsItem(this.value)
+    );
```
You're absolutely right. This is legacy code I missed to identify when adding the smart popovers.
I'll directly remove it. There is not much value in it.

Jinnie commented on src/clr-angular/forms/combobox/option.ts:
```diff
@@ -3,71 +3,94 @@
+    this.optionProxy.id = this._id;
+  }
+  get optionId() {
+    return this._id;
```
Duplicate comment, I guess.
Done, anyway.

gnomeontherun commented on src/clr-angular/forms/combobox/option-items.directive.ts:
```diff
@@ -0,0 +1,120 @@
+})
+export class ClrOptionItems<T> implements DoCheck, OnDestroy {
+  private iterableProxy: NgForOf<T>;
+  private _rawItems: T[];
```
You're right, I didn't see them since they weren't together. I thought eslint was supposed to require that.

gnomeontherun commented:
@chlane can you help us audit this one last time before we mark this as ready to release? Sample is at https://5ea800bea72f5baa27578765--vmware-clarity.netlify.app

erabl-identos commented:
I've had a look at the preview at https://5ea800bea72f5baa27578765--vmware-clarity.netlify.app/combobox and might have found a bug.

Steps to reproduce:
After loading the page the "_Favorite States (Async)_" example contains the two values _Arizona_ and _Arkansas_. Focus on the input and write a single letter. Removing the letter by pressing _backspace_ on the keyboard also removes the last selected value, in this case _Arkansas_.

Expected behaviour:
If one uses the _delete_ key instead the letter is removed without changing the selection. I would expect the _backspace_ to behave the same way and only remove the last selected value if there are no written letters afterwards.

Thank you for your great work!

clane commented:
This issues from testing are attached.

[clarity_050620combobox.xlsx](https://github.com/vmware/clarity/files/4590240/clarity_050620combobox.xlsx)

Jinnie commented:
@clane 
A11Y requests done, with exception of 1) and 2) which are common for all form elements and are not focus of this PR
Can be verified here:
https://5ec782484499f3fd69725086--vmware-clarity.netlify.app/combobox

PR is also updated to latest, monorepo, source tree. 
Website is **NOT** submitted and will be re-added after the ongoing website changes are finalized.

clane commented:
Can we add @ghadK to this review? I would like a second tester to review this is well

Jinnie commented:
@clane, there is nothing that could be stopping @gkhadK from reviewing this issue.
This is the working preview:
https://5ec782484499f3fd69725086--vmware-clarity.netlify.app/combobox
Contact me directly if any other help is needed.

chlane commented:
There are outstanding issue reported by @gkhadK. See 
[clarity_050620combobox.xlsx](https://github.com/vmware/clarity/files/4702714/clarity_050620combobox.xlsx)

Jinnie commented:
A11y status and questions. Issues starting with **(?)** need comment
(attn. @clane, @gkhadK)

> 1 The help text below the comboboxes lacks sufficient color contrast

Will Not Fix. This is common for all forms. If necessary - a new issue must be file (unless it already exists).

> 2 The helper text below the comboboxes is not read by VoiceOver

Fixed.

> 3 The label elements for the comboboxes is not read by VoiceOver

Fixed.

> 4 The popups associated with comboboxes should not appear on focus

Fixed

> 5 The combobox examples are still interactive after activating the Disabled Button

Fixed

> 6 Unnecessary screen reader only text

Fixed

> 7 Compact Combobox name="one" is missing a visible label

Will Not Fix. This is a test-case only, but we need to be sure it works. So, it is intentional for the dev app.

> 8

No 8 :)

> 9 The comboboxes with pills need a live region to communicate the addition or deletion of pills

As per Clarity design we won't be adding live regions in the framework. Users will do it on app-level. Added (clrSelectionChange) handler to the component. Should be enough to facilitate the user.

> 10 The error messages are not programmatically associated with their respective input fields

Fixed.
	
> 11	The meaning of the world, sun and home icons is not provided programmatically

This is not a bug, but in the user domain. Nevertheless, I updated the examples to follow your suggestions. So, it's fixed.
		
> 13 The autocomplete behavior should match against characters starting from the beginning of the string

**(?)** I’m confused about this. You are right that the W3C examples work like this (it is even easier), but our design specs request otherwise. It also makes sense if you're searching for "West Virginia" to be able to find it typing "virg..." not "wes.." only. And I can't find a specification that requests it, though the example works this way. We need to discuss it more.
		
> 14 Combobox: Required field info is not communicated to any users.

This is user domain - they provide the html here. No where in our forms examples we mark the required fields. Not in dev-app, not in the website.
But there is a note on the website, in the forms section that suggests using a mark (Optional) on the optional ones, as better approach.
(https://clarity.design/documentation/forms)
		
> 15 Grid layout combobox pills label is announced twice	

**(?)** Can not reproduce. I also added labels on the icons, so they are read better now, but I can’t reproduce this on old versions too.
		
> 16 Grid layout pills Combobox label is not announced and announced table details are incomplete when user navigates to that combobox using tab navigation

Working on this… not investigated fully yet
		
> 17 Pills List combobox: Table navigation doesn’t work with the nvda table keyboard shortcuts

This is a misunderstanding. Our grid has always 1 row, even when pills are many and wrap to the next line. So, we can’t expect that up and down arrows will be doing anything. We are following this grid example: https://www.w3.org/TR/wai-aria-practices/examples/grid/LayoutGrids.html (Example 2: Pill List...)
		
> 18 Already selected value in the combobox list is announced as Not selected when screen reader user navigates through the list.

**(?)** Here we have big difference between ARIA 1.0 and ARIA 1.1
From what I see, we’re following ARIA 1.1, but I have troubles with it when I want to use aria-selected. And I read in internet that 1.1 (for combobox) is not well supported yet in readers. Should we switch to 1.0 structure??
Still working on it and not tested it works better with 1.0
I use this article as reference:
https://www.levelaccess.com/differences-aria-1-0-1-1-changes-rolecombobox/

Jinnie commented:
@clane @gkhadK
Thanks for the great work on the a11y testing!
As we have all the issues in "Verified" state now, I will merge this PR.
In case some new issues or improvement opportunities are found, please create new issues - we will have enough time to address them in the forthcoming 4.0 beta and RC releases.
Thanks once again!
